packages/ts-transformers/src/closures/transformer.ts

```diff
@@ -0,0 +1,646 @@
+      // Recursively compare the object expressions
+      return expressionsMatch(a.expression, b.expression);
+    } else if (ts.isIdentifier(a) && ts.isIdentifier(b)) {
+      return a.text === b.text;
+    }
+    return false;
+  }
+
+  // Replace captured expressions with their parameter names
+  for (const [varName, capturedExpr] of captures) {
+    const visitor: ts.Visitor = (node) => {
+      // Check if this node matches the captured expression
+      if (ts.isExpression(node) && expressionsMatch(node, capturedExpr)) {
+        return factory.createIdentifier(varName);
+      }
+      return ts.visitEachChild(node, visitor, undefined);
+    };
+    transformedBody = ts.visitNode(
+      transformedBody,
+      visitor,
+    ) as typeof transformedBody;
+  }
+
+  // Create the new callback
+  const newCallback = factory.createArrowFunction(
+    callback.modifiers,
+    callback.typeParameters,
+    [destructuredParam],
+    callback.type,
+    ts.isArrowFunction(callback)
+      ? callback.equalsGreaterThanToken
+      : factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
+    transformedBody,
+  );
+
+  // Wrap in recipe() using the proper imported identifier
+  const recipeIdentifier = getHelperIdentifier(
+    factory,
+    context.sourceFile,
+    "recipe",
+  );
+  const recipeCall = factory.createCallExpression(
+    recipeIdentifier,
+    undefined,
+    [
+      factory.createStringLiteral("map with pattern including captures"),
+      newCallback,
+    ],
+  );
+
+  // Create the params object
+  const paramsObject = factory.createObjectLiteralExpression(paramProperties);
+
+  // Create map_with_pattern property access (e.g., state.items.map_with_pattern)
+  // mapCall.expression is a PropertyAccessExpression like state.items.map
+  // We need to replace "map" with "map_with_pattern"
+  const mapWithPatternAccess = ts.isPropertyAccessExpression(mapCall.expression)
+    ? factory.createPropertyAccessExpression(
+      mapCall.expression.expression, // state.items
+      factory.createIdentifier("map_with_pattern"),
+    )
+    : factory.createIdentifier("map_with_pattern"); // Fallback (shouldn't happen)
+
+  // Return the transformed map_with_pattern call with recipe as first arg, params as second arg
+  return factory.createCallExpression(
+    mapWithPatternAccess,
+    mapCall.typeArguments,
+    [recipeCall, paramsObject],
+  );
+}
+
+function transformClosures(context: TransformationContext): ts.SourceFile {
+  const { checker, factory, sourceFile } = context;
+
+  function visit(node: ts.Node): ts.Node {
+    // Check for OpaqueRef<T[]> or Cell<T[]> map calls with callbacks
+    if (ts.isCallExpression(node) && isOpaqueRefArrayMapCall(node, checker)) {
+      const callback = node.arguments[0];
+
+      // Check if the callback is an arrow function or function expression
+      if (
+        callback &&
+        (ts.isArrowFunction(callback) || ts.isFunctionExpression(callback))
```

Comment:
There's another similar helper function defined here `isFunctionDeclaration` that seems to include various types of function types -- if we don't care what *type* of function it is, maybe we can reuse that, or just add these utils to the generic ast helpers in `ast/` (i think that's the rough idea anyway)